packages/langium/src/grammar/workspace/grammar-workspace-manager.ts

```diff
@@ -0,0 +1,57 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import ignore, { Ignore } from 'ignore';
+import { CancellationToken, WorkspaceFolder } from 'vscode-languageserver-protocol';
+import { URI, Utils } from 'vscode-uri';
+import { LangiumSharedServices } from '../../services';
+import { ConfigurationProvider } from '../../workspace/configuration';
+import { FileSystemNode } from '../../workspace/file-system-provider';
+import { DefaultWorkspaceManager } from '../../workspace/workspace-manager';
+import { LangiumGrammarLanguageMetaData } from '../generated/module';
+
+const CONFIG_KEY = 'build';
+
+interface WorkspaceManagerConf {
+    /**
+     * gitignore separated exclusion patterns, separated by comma
+     */
+    ignorePatterns: string
+}
+
+export class LangiumGrammarWorkspaceManager extends DefaultWorkspaceManager {
+
+    protected readonly configurationProvider: ConfigurationProvider;
+    protected matcher: Ignore | null;
+
+    constructor(services: LangiumSharedServices) {
+        super(services);
+        this.configurationProvider = services.workspace.ConfigurationProvider;
+    }
+
+    async initializeWorkspace(folders: WorkspaceFolder[], cancelToken = CancellationToken.None): Promise<void> {
+        const buildConf: WorkspaceManagerConf = await this.configurationProvider.getConfiguration(LangiumGrammarLanguageMetaData.languageId, CONFIG_KEY);
+        const ignorePatterns = buildConf.ignorePatterns?.split(',')?.map(pattern => pattern.trim())?.filter(pattern => pattern.length > 0);
+        this.matcher = ignorePatterns ? ignore().add(ignorePatterns) : null;
+        return super.initializeWorkspace(folders, cancelToken);
+    }
+
+    protected includeEntry(entry: FileSystemNode, fileExtensions: string[]): boolean {
+        if (this.matcher) {
+            const parentWsFolder = this.folders?.map(folder => folder.uri).find(folderUri => entry.uri.toString().startsWith(folderUri));
+            let entryPath = entry.uri.path;
+            if (parentWsFolder) {
+                // create path relative to workspace folder root: /user/foo/workspace/entry.txt -> entry.txt
+                const wsFolderPath = URI.parse(parentWsFolder).path + '/';
+                entryPath = entryPath.replace(wsFolderPath, '');
```

Comment:
The string operations `startsWith` and `replace` seem unsafe to me: they can lead to false matches. Following my recommendation of moving this file to `sprotty-vscode`, you can use the Node.js `path` module for path operations.